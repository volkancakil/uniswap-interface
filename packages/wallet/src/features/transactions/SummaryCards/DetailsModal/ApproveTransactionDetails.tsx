import { SharedEventName } from '@uniswap/analytics-events'
import { useTranslation } from 'react-i18next'
import { Flex, Text, TouchableArea, isWeb } from 'ui/src'
import { iconSizes } from 'ui/src/theme'
import { CurrencyLogo } from 'uniswap/src/components/CurrencyLogo/CurrencyLogo'
import { useLocalizationContext } from 'uniswap/src/features/language/LocalizationContext'
import { ElementName, ModalName } from 'uniswap/src/features/telemetry/constants'
import { sendAnalyticsEvent } from 'uniswap/src/features/telemetry/send'
import { useCurrencyInfo } from 'uniswap/src/features/tokens/useCurrencyInfo'
import { ApproveTransactionInfo, TransactionDetails } from 'uniswap/src/features/transactions/types/transactionDetails'
import { getSymbolDisplayText } from 'uniswap/src/utils/currency'
import { NumberType } from 'utilities/src/format/types'
import { useWalletNavigation } from 'wallet/src/contexts/WalletNavigationContext'
import { buildCurrencyId } from 'wallet/src/utils/currencyId'

const INFINITE_AMOUNT = 'INF'
const ZERO_AMOUNT = '0.0'
export function ApproveTransactionDetails({
  transactionDetails,
  typeInfo,
  onClose,
}: {
  transactionDetails: TransactionDetails
  typeInfo: ApproveTransactionInfo
  onClose: () => void
}): JSX.Element {
  const { t } = useTranslation()
  const { formatNumberOrString } = useLocalizationContext()
  const { navigateToTokenDetails } = useWalletNavigation()
  const currencyInfo = useCurrencyInfo(buildCurrencyId(transactionDetails.chainId, typeInfo.tokenAddress))

  const { approvalAmount } = typeInfo

  const amount =
    approvalAmount === INFINITE_AMOUNT
      ? t('transaction.amount.unlimited')
      : approvalAmount && approvalAmount !== ZERO_AMOUNT
        ? formatNumberOrString({ value: approvalAmount, type: NumberType.TokenNonTx })
        : ''

  const symbol = getSymbolDisplayText(currencyInfo?.currency.symbol)

  const onPressToken = (): void => {
    if (currencyInfo) {
      sendAnalyticsEvent(SharedEventName.ELEMENT_CLICKED, {
        element: ElementName.TokenItem,
        modal: ModalName.TransactionDetails,
      })

      navigateToTokenDetails(currencyInfo.currencyId)
      if (!isWeb) {
        onClose()
      }
    }
  }

  return (
    <TouchableArea onPress={onPressToken}>
      <Flex centered gap="$spacing8" p="$spacing32">
        <Text variant="heading3">{amount}</Text>
        <Flex centered row gap="$spacing8">
          <CurrencyLogo currencyInfo={currencyInfo} size={iconSizes.icon20} />
          <Text color="$neutral2" variant="body2">
            {symbol}
          </Text>
        </Flex>
      </Flex>
    </TouchableArea>
  )
}
