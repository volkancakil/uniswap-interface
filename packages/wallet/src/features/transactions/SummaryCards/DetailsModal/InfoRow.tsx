import { PropsWithChildren } from 'react'
import { Flex } from 'ui/src'
import { ContentRow } from 'wallet/src/features/transactions/TransactionRequest/ContentRow'

export function InfoRow({
  label,
  children,
}: PropsWithChildren & {
  label: string
}): JSX.Element {
  return (
    <ContentRow label={label} variant="body3">
      <Flex centered row gap="$spacing4">
        {children}
      </Flex>
    </ContentRow>
  )
}
