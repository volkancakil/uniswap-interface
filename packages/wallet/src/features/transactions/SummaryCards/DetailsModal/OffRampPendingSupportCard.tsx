import { useTranslation } from 'react-i18next'
import { InlineCard, Text, TouchableArea } from 'ui/src'
import { ArrowUpRight, MessageQuestion } from 'ui/src/components/icons'
import { openOfframpPendingSupportLink } from 'wallet/src/utils/linking'

export function OffRampPendingSupportCard(): JSX.Element {
  const { t } = useTranslation()

  return (
    <TouchableArea p="$spacing8" onPress={openOfframpPendingSupportLink}>
      <InlineCard
        CtaButtonIcon={ArrowUpRight}
        CtaButtonIconColor="$neutral2"
        Icon={MessageQuestion}
        color="$neutral2"
        iconColor="$neutral2"
        description={
          <Text color="$neutral2" variant="body3">
            {t('transaction.status.sale.pendingCard.msg')}
          </Text>
        }
        heading={
          <Text color="$neutral1" variant="body3">
            {t('transaction.status.sale.pendingCard.title')}
          </Text>
        }
      />
    </TouchableArea>
  )
}
