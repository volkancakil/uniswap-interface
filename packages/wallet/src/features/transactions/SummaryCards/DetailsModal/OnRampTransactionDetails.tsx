import { useMemo } from 'react'
import { useLocalizationContext } from 'uniswap/src/features/language/LocalizationContext'
import { ValueType } from 'uniswap/src/features/tokens/getCurrencyAmount'
import { useCurrencyInfo } from 'uniswap/src/features/tokens/useCurrencyInfo'
import {
  OnRampPurchaseInfo,
  OnRampTransferInfo,
  TransactionDetails,
  TransactionType,
} from 'uniswap/src/features/transactions/types/transactionDetails'
import { getSymbolDisplayText } from 'uniswap/src/utils/currency'
import { NumberType } from 'utilities/src/format/types'
import { CurrencyTransferContent } from 'wallet/src/features/transactions/SummaryCards/DetailsModal/TransferTransactionDetails'
import { isOnRampPurchaseTransactionInfo } from 'wallet/src/features/transactions/SummaryCards/DetailsModal/types'
import { useFormattedCurrencyAmountAndUSDValue } from 'wallet/src/features/transactions/SummaryCards/DetailsModal/utils'
import { buildCurrencyId } from 'wallet/src/utils/currencyId'

export function OnRampTransactionDetails({
  transactionDetails,
  typeInfo,
  onClose,
}: {
  transactionDetails: TransactionDetails
  typeInfo: OnRampTransferInfo | OnRampPurchaseInfo
  onClose: () => void
}): JSX.Element {
  const formatter = useLocalizationContext()
  const currencyInfo = useCurrencyInfo(buildCurrencyId(transactionDetails.chainId, typeInfo.destinationTokenAddress))

  const { amount, value: calculatedValue } = useFormattedCurrencyAmountAndUSDValue({
    currency: currencyInfo?.currency,
    currencyAmountRaw: typeInfo.destinationTokenAmount?.toString(),
    formatter,
    isApproximateAmount: false,
    valueType: ValueType.Exact,
  })

  const transactionValue = useMemo(() => {
    return isOnRampPurchaseTransactionInfo(typeInfo)
      ? formatter.formatNumberOrString({
          value: typeInfo.sourceAmount,
          type: NumberType.FiatTokenPrice,
          currencyCode: typeInfo.sourceCurrency,
        })
      : calculatedValue
  }, [typeInfo, formatter, calculatedValue])

  const symbol = getSymbolDisplayText(currencyInfo?.currency.symbol)

  const tokenAmountWithSymbol = symbol ? amount + ' ' + symbol : amount // Prevents 'undefined' from being displayed

  return (
    <CurrencyTransferContent
      currencyInfo={currencyInfo}
      showValueAsHeading={typeInfo.type === TransactionType.OnRampPurchase}
      tokenAmountWithSymbol={tokenAmountWithSymbol}
      value={transactionValue}
      onClose={onClose}
    />
  )
}
