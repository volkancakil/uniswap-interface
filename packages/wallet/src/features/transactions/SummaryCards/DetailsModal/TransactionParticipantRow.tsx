import { PlatformSplitStubError } from 'utilities/src/errors'
import { TransactionParticipantRowProps } from 'wallet/src/features/transactions/SummaryCards/DetailsModal/types'

export function TransactionParticipantRow(_props: TransactionParticipantRowProps): JSX.Element {
  throw new PlatformSplitStubError('TransactionParticipantRow')
}
