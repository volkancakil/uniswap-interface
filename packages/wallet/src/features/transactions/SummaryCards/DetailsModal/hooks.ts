import { SharedEventName } from '@uniswap/analytics-events'
import { useCallback } from 'react'
import { isWeb } from 'ui/src'
import { CurrencyInfo } from 'uniswap/src/features/dataApi/types'
import { useLocalizationContext } from 'uniswap/src/features/language/LocalizationContext'
import { ElementName, ModalName } from 'uniswap/src/features/telemetry/constants'
import { sendAnalyticsEvent } from 'uniswap/src/features/telemetry/send'
import { ValueType } from 'uniswap/src/features/tokens/getCurrencyAmount'
import { useCurrencyInfo } from 'uniswap/src/features/tokens/useCurrencyInfo'
import { isUniswapX } from 'uniswap/src/features/transactions/swap/utils/routing'
import { TransactionDetails, isFinalizedTx } from 'uniswap/src/features/transactions/types/transactionDetails'
import { useWalletNavigation } from 'wallet/src/contexts/WalletNavigationContext'
import { useFormattedCurrencyAmountAndUSDValue } from 'wallet/src/features/transactions/SummaryCards/DetailsModal/utils'
import { buildCurrencyId, buildNativeCurrencyId } from 'wallet/src/utils/currencyId'

export function useNetworkFee(transactionDetails: TransactionDetails): {
  value: string
  amount: string
} {
  const formatter = useLocalizationContext()

  const currencyId = transactionDetails.networkFee
    ? buildCurrencyId(transactionDetails.chainId, transactionDetails.networkFee.tokenAddress)
    : buildNativeCurrencyId(transactionDetails.chainId)
  const currencyInfo = useCurrencyInfo(currencyId)

  const currencyAmountRaw =
    transactionDetails.networkFee?.quantity != null
      ? transactionDetails.networkFee.quantity
      : isFinalizedTx(transactionDetails)
        ? '0'
        : undefined

  return useFormattedCurrencyAmountAndUSDValue({
    currency: currencyInfo?.currency,
    currencyAmountRaw,
    valueType: ValueType.Exact,
    formatter,
    isApproximateAmount: false,
    isUniswapX: isUniswapX(transactionDetails),
  })
}

export function useTokenDetailsNavigation(currency: Maybe<CurrencyInfo>, onClose?: () => void): () => void {
  const { navigateToTokenDetails } = useWalletNavigation()

  return useCallback(() => {
    if (currency) {
      sendAnalyticsEvent(SharedEventName.ELEMENT_CLICKED, {
        element: ElementName.TokenItem,
        modal: ModalName.TransactionDetails,
      })

      navigateToTokenDetails(currency.currencyId)
      if (!isWeb) {
        onClose?.()
      }
    }
  }, [currency, navigateToTokenDetails, onClose])
}
