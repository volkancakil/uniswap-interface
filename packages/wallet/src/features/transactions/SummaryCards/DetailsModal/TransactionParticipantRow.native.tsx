import { useMemo } from 'react'
import { useTranslation } from 'react-i18next'
import ContextMenu from 'react-native-context-menu-view'
import { useDispatch } from 'react-redux'
import { pushNotification } from 'uniswap/src/features/notifications/slice'
import { AppNotificationType, CopyNotificationType } from 'uniswap/src/features/notifications/types'
import { setClipboard } from 'uniswap/src/utils/clipboard'
import { useWalletNavigation } from 'wallet/src/contexts/WalletNavigationContext'
import { InfoRow } from 'wallet/src/features/transactions/SummaryCards/DetailsModal/InfoRow'
import { TransactionParticipantDisplay } from 'wallet/src/features/transactions/SummaryCards/DetailsModal/TransactionParticipantDisplay'
import { TransactionParticipantRowProps } from 'wallet/src/features/transactions/SummaryCards/DetailsModal/types'

export function TransactionParticipantRow({
  onClose,
  address,
  isSend = false,
}: TransactionParticipantRowProps): JSX.Element {
  const { t } = useTranslation()
  const dispatch = useDispatch()
  const { navigateToExternalProfile } = useWalletNavigation()

  const onCopyAddress = (): void => {
    setClipboard(address)
      .then(() => {
        dispatch(
          pushNotification({
            type: AppNotificationType.Copied,
            copyType: CopyNotificationType.Address,
          }),
        )
      })
      .catch(() => {
        // setClipboard shouldn't ever error
      })
  }

  const onViewProfile = (): void => {
    navigateToExternalProfile({ address })
    onClose()
  }

  const menuActions = useMemo(() => {
    return [{ title: t('common.copy.address') }, { title: t('common.view.profile') }]
  }, [t])

  return (
    <InfoRow label={isSend ? t('common.text.recipient') : t('common.text.sender')}>
      <ContextMenu
        dropdownMenuMode
        actions={menuActions}
        onPress={(e): void => {
          // Emitted index based on order of menu action array
          // Copy address
          if (e.nativeEvent.index === 0) {
            onCopyAddress()
          }
          // View profile
          if (e.nativeEvent.index === 1) {
            onViewProfile()
          }
        }}
      >
        <TransactionParticipantDisplay address={address} />
      </ContextMenu>
    </InfoRow>
  )
}
