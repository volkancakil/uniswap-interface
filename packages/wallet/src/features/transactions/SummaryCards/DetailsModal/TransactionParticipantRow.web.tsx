import { useTranslation } from 'react-i18next'
import { useDispatch } from 'react-redux'
import { Flex, Popover } from 'ui/src'
import { CopyAlt } from 'ui/src/components/icons'
import { pushNotification } from 'uniswap/src/features/notifications/slice'
import { AppNotificationType, CopyNotificationType } from 'uniswap/src/features/notifications/types'
import { setClipboard } from 'uniswap/src/utils/clipboard'
import { isMobileApp } from 'utilities/src/platform'
import { MenuContent } from 'wallet/src/components/menu/MenuContent'
import { MenuContentItem } from 'wallet/src/components/menu/types'
import { InfoRow } from 'wallet/src/features/transactions/SummaryCards/DetailsModal/InfoRow'
import { TransactionParticipantDisplay } from 'wallet/src/features/transactions/SummaryCards/DetailsModal/TransactionParticipantDisplay'
import { TransactionParticipantRowProps } from 'wallet/src/features/transactions/SummaryCards/DetailsModal/types'

export function TransactionParticipantRow({ address, isSend = false }: TransactionParticipantRowProps): JSX.Element {
  const { t } = useTranslation()
  const dispatch = useDispatch()

  const onCopyAddress = async (): Promise<void> => {
    await setClipboard(address)
    dispatch(
      pushNotification({
        type: AppNotificationType.Copied,
        copyType: CopyNotificationType.Address,
      }),
    )
  }

  const options: MenuContentItem[] = [
    {
      label: t('common.copy.address'),
      onPress: onCopyAddress,
      Icon: !isMobileApp ? CopyAlt : undefined,
      iconPlacement: 'left',
      iconTextGap: '$spacing8',
      iconProps: {
        size: '$icon.16',
      },
      textProps: {
        variant: 'body4',
        pr: '$spacing16',
      },
    },
  ]

  return (
    <InfoRow label={isSend ? t('common.text.recipient') : t('common.text.sender')}>
      <Popover hoverable placement="top-end" offset={{ mainAxis: 4, crossAxis: 8 }}>
        <Popover.Trigger>
          <TransactionParticipantDisplay address={address} />
        </Popover.Trigger>
        <Popover.Content
          animation={[
            'quick',
            {
              opacity: {
                overshootClamping: true,
              },
            },
          ]}
          borderColor="$surface3"
          borderRadius="$rounded12"
          borderWidth="$spacing1"
          disableRemoveScroll={false}
          enterStyle={{ y: -10, opacity: 0 }}
          exitStyle={{ y: -10, opacity: 0 }}
          p="$none"
        >
          <Flex>
            <MenuContent items={options} p="$spacing4" borderRadius="$rounded12" />
          </Flex>
        </Popover.Content>
      </Popover>
    </InfoRow>
  )
}
