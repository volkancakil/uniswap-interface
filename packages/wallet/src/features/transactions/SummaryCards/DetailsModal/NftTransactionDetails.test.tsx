import { NFTMintTransactionInfo, TransactionDetails } from 'uniswap/src/features/transactions/types/transactionDetails'
import { NftTransactionDetails } from 'wallet/src/features/transactions/SummaryCards/DetailsModal/NftTransactionDetails'
import { ACCOUNT, preloadedWalletPackageState } from 'wallet/src/test/fixtures'
import { render } from 'wallet/src/test/test-utils'

const preloadedState = preloadedWalletPackageState({ account: ACCOUNT })
const nftTypeInfo = {
  type: 'nft-mint',
  nftSummaryInfo: {
    tokenId: '9920dbad-ff24-47c8-814a-094566fc45ff',
    name: 'voluptates repudiandae aliquid',
    collectionName: 'inventore qui fugiat',
    imageURL: 'https://loremflickr.com/640/480',
    address: '0xaea14f6cccfeae34fea11d9a2ca6aabb112e8b8d',
  },
} as NFTMintTransactionInfo
const mockTransaction = {
  id: '9920dbad-ff24-47c8-814a-094566fc45ff',
  chainId: 81457,
  routing: 'CLASSIC',
  from: '0xee814caea14f6cccfeae34fea11d9a2ca6aabb11',
  typeInfo: nftTypeInfo,
  status: 'confirmed',
  addedTime: 1719911758204,
  options: { request: {} },
  hash: 'b568a9e9-bbe7-42fc-ab00-5070186c0600',
  receipt: {
    transactionIndex: 29529,
    blockNumber: 17489,
    blockHash: 'dfd3ad45-78e7-4124-90f2-92758b4499ba',
    confirmedTime: 1719946653408,
    confirmations: 57408,
    gasUsed: 27844,
    effectiveGasPrice: 2941,
  },
} as TransactionDetails

// Mock the ImageUri component
jest.mock('wallet/src/features/images/ImageUri', () => ({
  ImageUri: jest.fn(() => null),
}))

describe('NftTransactionDetails Component', () => {
  it('renders NftTransactionDetails without error', () => {
    const onClose = jest.fn()

    const tree = render(
      <NftTransactionDetails transactionDetails={mockTransaction} typeInfo={nftTypeInfo} onClose={onClose} />,
      {
        preloadedState,
      },
    )

    expect(tree).toMatchSnapshot()
  })
})
