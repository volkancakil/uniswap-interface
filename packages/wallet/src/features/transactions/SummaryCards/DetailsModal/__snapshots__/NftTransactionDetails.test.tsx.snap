// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NftTransactionDetails Component renders NftTransactionDetails without error 1`] = `
<View
  style={
    {
      "borderBottomLeftRadius": 20,
      "borderBottomRightRadius": 20,
      "borderTopLeftRadius": 20,
      "borderTopRightRadius": 20,
      "flexDirection": "column",
      "overflow": "hidden",
    }
  }
>
  <View
    cancelable={true}
    disabled={false}
    focusable={true}
    hitSlop={[Function]}
    minPressDuration={0}
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onPress={[Function]}
    onPressIn={[Function]}
    onPressOut={[Function]}
    style={
      {
        "flexDirection": "column",
        "opacity": 1,
        "transform": [
          {
            "scale": 1,
          },
        ],
      }
    }
  />
  <View
    style={
      {
        "borderBottomColor": "rgba(34,34,34,0.05)",
        "borderBottomLeftRadius": 20,
        "borderBottomRightRadius": 20,
        "borderBottomWidth": 1,
        "borderLeftColor": "rgba(34,34,34,0.05)",
        "borderLeftWidth": 1,
        "borderRightColor": "rgba(34,34,34,0.05)",
        "borderRightWidth": 1,
        "borderStyle": "solid",
        "borderTopColor": "rgba(34,34,34,0.05)",
        "borderTopWidth": 0,
        "flexDirection": "column",
        "paddingBottom": 12,
        "paddingLeft": 12,
        "paddingRight": 12,
        "paddingTop": 12,
      }
    }
  >
    <Text
      allowFontScaling={false}
      maxFontSizeMultiplier={1.4}
      style={
        {
          "color": "#222222",
          "fontFamily": "Basel Grotesk",
          "fontSize": 16,
          "fontWeight": "400",
          "lineHeight": 24,
        }
      }
      suppressHighlighting={true}
    >
      voluptates repudiandae aliquid
    </Text>
    <View
      disabled={true}
      hitSlop={
        {
          "bottom": 5,
          "left": 5,
          "right": 5,
          "top": 5,
        }
      }
      style={
        {
          "flexDirection": "column",
          "opacity": 1,
          "transform": [
            {
              "scale": 1,
            },
          ],
        }
      }
    >
      <View
        style={
          {
            "flexDirection": "row",
            "paddingRight": 12,
          }
        }
      >
        <Text
          allowFontScaling={false}
          maxFontSizeMultiplier={1.4}
          numberOfLines={1}
          style={
            {
              "color": "#7D7D7D",
              "fontFamily": "Basel Grotesk",
              "fontSize": 14,
              "fontWeight": "400",
              "lineHeight": 20,
            }
          }
          suppressHighlighting={true}
        >
          inventore qui fugiat
        </Text>
      </View>
    </View>
  </View>
</View>
`;
